src/main/java/com/epam/aidial/core/controller/DownloadFileController.java

```diff
@@ -29,40 +29,38 @@ protected Future<?> handle(ResourceDescription resource, boolean hasWriteAccess)
+                    }
 
-        Promise<Void> result = Promise.promise();
-        blobFuture.onSuccess(blob -> {
-            if (blob == null) {
-                context.respond(HttpStatus.NOT_FOUND);
-                result.complete();
-                return;
-            }
+                    Payload payload = blob.getPayload();
+                    MutableContentMetadata metadata = payload.getContentMetadata();
+                    String contentType = metadata.getContentType();
+                    Long length = metadata.getContentLength();
 
-            Payload payload = blob.getPayload();
-            MutableContentMetadata metadata = payload.getContentMetadata();
-            String contentType = metadata.getContentType();
-            Long length = metadata.getContentLength();
+                    HttpServerResponse response = context.getResponse()
+                            .putHeader(HttpHeaders.CONTENT_TYPE, contentType)
+                            // content-length removed by vertx
+                            .putHeader(HttpHeaders.CONTENT_LENGTH, length.toString());
 
-            HttpServerResponse response = context.getResponse()
-                    .putHeader(HttpHeaders.CONTENT_TYPE, contentType)
-                    // content-length removed by vertx
-                    .putHeader(HttpHeaders.CONTENT_LENGTH, length.toString());
+                    try {
+                        InputStreamReader stream = new InputStreamReader(proxy.getVertx(), payload.openStream());
+                        stream.pipeTo(response)
+                                .onFailure(error -> {
+                                    stream.close();
+                                    context.getResponse().reset();
+                                });
 
-            try {
-                InputStreamReader stream = new InputStreamReader(proxy.getVertx(), payload.openStream());
-                stream.pipeTo(response, result);
-                result.future().onFailure(error -> {
-                    stream.close();
-                    context.getResponse().reset();
-                });
-            } catch (IOException e) {
-                result.fail(e);
-            }
-        }).onFailure(error -> context.respond(HttpStatus.INTERNAL_SERVER_ERROR,
-                "Failed to fetch file with path %s/%s".formatted(resource.getBucketName(), resource.getOriginalPath())));
+                        return Future.succeededFuture();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }).onFailure(error -> context.respond(HttpStatus.INTERNAL_SERVER_ERROR,
+                        "Failed to fetch file with path %s/%s".formatted(resource.getBucketName(),
+                                resource.getOriginalPath())));
 
-        return result.future();
+        return Future.succeededFuture();
```

Comment:
same here, previous implementation succeed future only after sending file content to the response

Reply:
Check the contract of handle.